File: WinForms/CS/AIMainDemo/Modules/SmartPasteAIModule.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Runtime.CompilerServices;
using System.Windows.Forms;
using DevExpress.AIIntegration.WinForms;
using DevExpress.Data.Utils;
using DevExpress.DXperience.Demos;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraLayout;

namespace DevExpress.AI.Demos {
    public partial class SmartPasteAIModule : AITutorialControl {
        private const string LayoutCopyText = @"Payment: Amount - $123.00
Statement Date: 10/15/2024
Account Number: 133244556
Name: John Smith
Contact: (123) 456-7890
Email: john@myemail.com
Address:
- 123 Elm St Apt 4B
- New York, NY 10001";
        private const string GridCopyText = @"Transactions 
- Hanari Carnes - Invoice amount due: $123.00, due by 10/30/2024. Status: Processed ✅
- Que Delicia - Amount: $231.00 | Due date: 10/29/2024 | Not processed yet 
- Romero y tomillo - $369.00 required by 10/28/2024, Processed ✅ 
- Mere Paillard - Payment: $158.00, due: 10/31/2024, unprocessed 
- Eastern Delights has an outstanding balance of $97.00, which was due on 10/27/2024. Status: Processed ✅ ";
        private const string SchedulerCopyText = @"Weekly Meeting with Mike F. from the Marketing Department
Date & Time: Tomorrow at 3:30 PM (New York time)
Duration: 45 minutes
Location: Conference room, 33rd floor, HQ office
Topic: Finalizing the marketing budget for the new promo campaign

Agenda: 

- Share what was completed
- Discuss issues
- Review incomplete tasks/requirements 

Reminder: 1 hour before the meeting.";
        GridColumn colCompany;
        GridColumn colID;
        GridColumn colDate;
        GridColumn colPayment;
        GridColumn colProcessed;

        public SmartPasteAIModule() {
            InitializeComponent();
            InitLayoutControl();
            tabPane1_SelectedPageIndexChanged(tabPane1, EventArgs.Empty);
            UpdatePaddings();
            schedulerDataStorage1.Resources.Add(1, "Meeting Room \"Falcon\"");
            schedulerDataStorage1.Resources.Add(2, "Conference room");
        }

        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            gridControl1.ForceInitialize();
            InitGridControl();
        }

        protected override void DoHide() {
            behaviorManager1.GetBehavior<SmartPasteBehavior>(layoutControl1)?.CloseOverlayForm();
            behaviorManager1.GetBehavior<SmartPasteBehavior>(gridView1)?.CloseOverlayForm();
            behaviorManager1.GetBehavior<SmartPasteSchedulerBehavior>(schedulerControl1)?.CloseOverlayForm();
            base.DoHide();
        }

        protected override void UpdatePaddings() {
            UpdateBottomSpacings(lciUcDisclaimerPanel1, ucDisclaimerPanel1.RoundedPanelOffset);
        }

        void InitLayoutControl() {
            foreach(BaseLayoutItem item in layoutControl1.Items)
                if(item is LayoutControlItem lci && lci.Control is TextEdit control)
                    control.KeyDown += LayoutControlProcessKeyDown;
            //
            var itemsDescriptions = new List<AIItemDescription>() {
                new AIItemDescription(layoutControlItem1, "The primary address line, which usually includes the street number and name (e.g., 123 Main St). Can include alphabetic and numeric characters."),
                new AIItemDescription(layoutControlItem2, "The secondary address line, which can include apartment or suite number (e.g., Apt 4B). This field is optional. Can include alphabetic and numeric characters."),
                new AIItemDescription(layoutControlItem3, "The city name. Only alphabetic characters are expected."),
                new AIItemDescription(layoutControlItem4, "The state, province, or region name or abbreviation. Only alphabetic characters are expected."),
                new AIItemDescription(layoutControlItem5, "The ZIP or postal code. Numeric characters and sometimes alphabetic characters (e.g., for Canadian postal codes) are expected."),
                new AIItemDescription(layoutControlItem6, "The first name of the recipient. Only alphabetic characters are expected."),
                new AIItemDescription(layoutControlItem7, "The last name of the recipient. Only alphabetic characters are expected."),
                new AIItemDescription(layoutControlItem8, "The phone number of the recipient in the format of (XXX) XXX-XXXX or similar. Only numeric characters and standard phone number formatting characters (e.g., parentheses, dashes, spaces) are expected."),
                new AIItemDescription(layoutControlItem10, "The email of the recipient.")
            };
            //
            var smartPasteBehavior = behaviorManager1.GetBehavior<SmartPasteBehavior>(layoutControl1);
            smartPasteBehavior.Properties.ItemDescriptions.AddRange(itemsDescriptions);
            layoutControl1.AllowCustomization = false;
        }

        void InitGridControl() {
            gridControl1.DataSource = SmartPasteDataHelper.GetData(10);
            // Obtain created columns.
            colCompany = gridView1.Columns["CompanyName"];
            colCompany.Caption = "Company Name";
            colID = gridView1.Columns["ID"];
            colDate = gridView1.Columns["DueDate"];
            colDate.Caption = "Due Date";
            colPayment = gridView1.Columns["Amount"];
            colPayment.Caption = "Amount";
            colProcessed = gridView1.Columns["Processed"];
            colProcessed.Caption = "Processed";

            // The Company column uses a ComboBox in-place editor that shows a list of available companies.
            RepositoryItemComboBox riComboBox = new RepositoryItemComboBox();
            riComboBox.Items.AddRange(SmartPasteDataHelper.companies);
            gridControl1.RepositoryItems.Add(riComboBox);
            colCompany.ColumnEdit = riComboBox;

            // Hide a column.
            colID.Visible = false;

            gridControl1.MainView.InvalidValueException += ViewInvalidValueException;
            gridControl1.ProcessGridKey += GridControlProcessKeyDown;
            (gridControl1.MainView as GridView).FocusedRowHandle = 0;

            var itemsDescriptions = new List<AIItemDescription>() {
                new AIItemDescription(colPayment, "The total monetary value of the order, typically in USD."),
                new AIItemDescription(colDate, "The date by which the company expects the order to be fulfilled."),
                new AIItemDescription(colProcessed, "A Boolean value indicating whether the order has been processed.")
            };
            var smartPasteBehavior = behaviorManager1.GetBehavior<SmartPasteBehavior>(gridControl1.MainView);
            smartPasteBehavior.Properties.ItemDescriptions.AddRange(itemsDescriptions);
        }
        void ViewInvalidValueException(object sender, InvalidValueExceptionEventArgs e) {
            e.ExceptionMode = DevExpress.XtraEditors.Controls.ExceptionMode.Ignore;
        }
        async void GridControlProcessKeyDown(object sender, KeyEventArgs e) {
            if(e.KeyData == (Keys.Control | Keys.Shift | Keys.V)) {
                e.Handled = true;
                var smartPasteBehavior = behaviorManager1.GetBehavior<SmartPasteBehavior>(gridControl1.MainView);
                await smartPasteBehavior.SmartPasteAsync();
            }
        }
        async void LayoutControlProcessKeyDown(object sender, KeyEventArgs e) {
            if(e.KeyData == (Keys.Control | Keys.Shift | Keys.V)) {
                e.Handled = true;
                var smartPasteBehavior = behaviorManager1.GetBehavior<SmartPasteBehavior>(layoutControl1);
                await smartPasteBehavior.SmartPasteAsync();
            }
        }
        void simpleButtonCopy_Click(object sender, EventArgs e) {
            SafeClipboardWin.Instance.SetText(memoEdit1.Text);
        }
        async void simpleButtonPaste_Click(object sender, EventArgs e) {
            var smartPasteBehavior = behaviorManager1.GetBehavior<SmartPasteBehavior>(layoutControl1);
            await smartPasteBehavior.SmartPasteAsync();
        }
        void tabPane1_SelectedPageIndexChanged(object sender, EventArgs e) {
            if(tabPane1.SelectedPage == layoutTabNavigationPage) {
                memoEdit1.Text = LayoutCopyText;
            }
            else if(tabPane1.SelectedPage == gridTabNavigationPage) {
                memoEdit1.Text = GridCopyText;
            }
            else {
                memoEdit1.Text = SchedulerCopyText;
            }            
        }
    }

    public class Record : INotifyPropertyChanged {
        public Record() { }
        int id;
        public int ID {
            get { return id; }
            set {
                if(id != value) {
                    id = value;
                    OnPropertyChanged();
                }
            }
        }
        string text;
        [DisplayName("Company")]
        public string CompanyName {
            get { return text; }
            set {
                if(text != value) {
                    if(string.IsNullOrEmpty(value))
                        throw new Exception();
                    text = value;
                    OnPropertyChanged();
                }
            }
        }
        decimal? val;
        [DataType(System.ComponentModel.DataAnnotations.DataType.Currency)]
        [DisplayName("Payment")]
        public decimal? Amount {
            get { return val; }
            set {
                if(val != value) {
                    val = value;
                    OnPropertyChanged();
                }
            }
        }
        DateTime dt;
        [DisplayFormat(DataFormatString = "d")]
        public DateTime DueDate {
            get { return dt; }
            set {
                if(dt != value) {
                    dt = value;
                    OnPropertyChanged();
                }
            }
        }
        bool state;
        public bool Processed {
            get { return state; }
            set {
                if(state != value) {
                    state = value;
                    OnPropertyChanged();
                }
            }
        }
        public override string ToString() {
            return string.Format("ID = {0}, Text = {1}", ID, CompanyName);
        }
        public event PropertyChangedEventHandler PropertyChanged;
        protected void OnPropertyChanged([CallerMemberName] string propertyName = "") {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }
    }

    public class SmartPasteDataHelper {
        public static string[] companies = new string[] { "Hanari Carnes", "Que Delicia", "Romero y tomillo", "Mere Paillarde",
            "Comércio Mineiro", "Reggiani Caseifici", "Maison Dewey", "Eastern Delights" };
        public static string[] items = new string[] { "Office Chairs", "Laptops", "Printer Ink", "Tables",
            "Paper" };
        public static BindingList<Record> GetData(int count) {
            BindingList<Record> records = new BindingList<Record>();
            var rnd = DevExpress.Data.Utils.NonCryptographicRandom.Default;
            for(int i = 0; i < count; i++) {
                int n = rnd.Next(10);
                var record = new Record() {
                    ID = i + 100,
                    CompanyName = companies[i % companies.Length],
                    DueDate = DateTime.Today.AddDays(n - 5),
                    Amount = i % 2 == 0 ? (i + 1) * 123 : i * 231,
                    Processed = i % 2 == 0,
                };
                records.Add(record);
            };
            return records;
        }
    }
}

File: WinForms/VB/AIMainDemo/Modules/SmartPasteAIModule.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.ComponentModel.DataAnnotations
Imports System.Runtime.CompilerServices
Imports System.Windows.Forms
Imports DevExpress.AIIntegration.WinForms
Imports DevExpress.Data.Utils
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraLayout

Namespace DevExpress.AI.Demos

    Public Partial Class SmartPasteAIModule
        Inherits AITutorialControl

        Private Const LayoutCopyText As String = "Payment: Amount - $123.00
Statement Date: 10/15/2024
Account Number: 133244556
Name: John Smith
Contact: (123) 456-7890
Email: john@myemail.com
Address:
- 123 Elm St Apt 4B
- New York, NY 10001"

        Private Const GridCopyText As String = "Transactions 
- Hanari Carnes - Invoice amount due: $123.00, due by 10/30/2024. Status: Processed ✅
- Que Delicia - Amount: $231.00 | Due date: 10/29/2024 | Not processed yet 
- Romero y tomillo - $369.00 required by 10/28/2024, Processed ✅ 
- Mere Paillard - Payment: $158.00, due: 10/31/2024, unprocessed 
- Eastern Delights has an outstanding balance of $97.00, which was due on 10/27/2024. Status: Processed ✅ "

        Private Const SchedulerCopyText As String = "Weekly Meeting with Mike F. from the Marketing Department
Date & Time: Tomorrow at 3:30 PM (New York time)
Duration: 45 minutes
Location: Conference room, 33rd floor, HQ office
Topic: Finalizing the marketing budget for the new promo campaign

Agenda: 

- Share what was completed
- Discuss issues
- Review incomplete tasks/requirements 

Reminder: 1 hour before the meeting."

        Private colCompany As GridColumn

        Private colID As GridColumn

        Private colDate As GridColumn

        Private colPayment As GridColumn

        Private colProcessed As GridColumn

        Public Sub New()
            InitializeComponent()
            InitLayoutControl()
            tabPane1_SelectedPageIndexChanged(tabPane1, EventArgs.Empty)
            UpdatePaddings()
            schedulerDataStorage1.Resources.Add(1, "Meeting Room ""Falcon""")
            schedulerDataStorage1.Resources.Add(2, "Conference room")
        End Sub

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            gridControl1.ForceInitialize()
            InitGridControl()
        End Sub

        Protected Overrides Sub DoHide()
            behaviorManager1.GetBehavior(Of SmartPasteBehavior)(layoutControl1)?.CloseOverlayForm()
            behaviorManager1.GetBehavior(Of SmartPasteBehavior)(gridView1)?.CloseOverlayForm()
            behaviorManager1.GetBehavior(Of SmartPasteSchedulerBehavior)(schedulerControl1)?.CloseOverlayForm()
            MyBase.DoHide()
        End Sub

        Protected Overrides Sub UpdatePaddings()
            UpdateBottomSpacings(lciUcDisclaimerPanel1, ucDisclaimerPanel1.RoundedPanelOffset)
        End Sub

        Private Sub InitLayoutControl()
            Dim lci As LayoutControlItem = Nothing, control As TextEdit = Nothing
            For Each item As BaseLayoutItem In layoutControl1.Items
                If CSharpImpl.__Assign(lci, TryCast(item, LayoutControlItem)) IsNot Nothing AndAlso CSharpImpl.__Assign(control, TryCast(lci.Control, TextEdit)) IsNot Nothing Then AddHandler control.KeyDown, AddressOf Me.LayoutControlProcessKeyDown
            Next

            '
            Dim itemsDescriptions = New List(Of AIItemDescription)() From {New AIItemDescription(layoutControlItem1, "The primary address line, which usually includes the street number and name (e.g., 123 Main St). Can include alphabetic and numeric characters."), New AIItemDescription(layoutControlItem2, "The secondary address line, which can include apartment or suite number (e.g., Apt 4B). This field is optional. Can include alphabetic and numeric characters."), New AIItemDescription(layoutControlItem3, "The city name. Only alphabetic characters are expected."), New AIItemDescription(layoutControlItem4, "The state, province, or region name or abbreviation. Only alphabetic characters are expected."), New AIItemDescription(layoutControlItem5, "The ZIP or postal code. Numeric characters and sometimes alphabetic characters (e.g., for Canadian postal codes) are expected."), New AIItemDescription(layoutControlItem6, "The first name of the recipient. Only alphabetic characters are expected."), New AIItemDescription(layoutControlItem7, "The last name of the recipient. Only alphabetic characters are expected."), New AIItemDescription(layoutControlItem8, "The phone number of the recipient in the format of (XXX) XXX-XXXX or similar. Only numeric characters and standard phone number formatting characters (e.g., parentheses, dashes, spaces) are expected."), New AIItemDescription(layoutControlItem10, "The email of the recipient.")}
            '
            Dim smartPasteBehavior = behaviorManager1.GetBehavior(Of SmartPasteBehavior)(layoutControl1)
            smartPasteBehavior.Properties.ItemDescriptions.AddRange(itemsDescriptions)
            layoutControl1.AllowCustomization = False
        End Sub

        Private Sub InitGridControl()
            gridControl1.DataSource = SmartPasteDataHelper.GetData(10)
            ' Obtain created columns.
            colCompany = gridView1.Columns("CompanyName")
            colCompany.Caption = "Company Name"
            colID = gridView1.Columns("ID")
            colDate = gridView1.Columns("DueDate")
            colDate.Caption = "Due Date"
            colPayment = gridView1.Columns("Amount")
            colPayment.Caption = "Amount"
            colProcessed = gridView1.Columns("Processed")
            colProcessed.Caption = "Processed"
            ' The Company column uses a ComboBox in-place editor that shows a list of available companies.
            Dim riComboBox As RepositoryItemComboBox = New RepositoryItemComboBox()
            riComboBox.Items.AddRange(SmartPasteDataHelper.companies)
            gridControl1.RepositoryItems.Add(riComboBox)
            colCompany.ColumnEdit = riComboBox
            ' Hide a column.
            colID.Visible = False
            AddHandler gridControl1.MainView.InvalidValueException, AddressOf ViewInvalidValueException
            AddHandler gridControl1.ProcessGridKey, AddressOf GridControlProcessKeyDown
            TryCast(gridControl1.MainView, GridView).FocusedRowHandle = 0
            Dim itemsDescriptions = New List(Of AIItemDescription)() From {New AIItemDescription(colPayment, "The total monetary value of the order, typically in USD."), New AIItemDescription(colDate, "The date by which the company expects the order to be fulfilled."), New AIItemDescription(colProcessed, "A Boolean value indicating whether the order has been processed.")}
            Dim smartPasteBehavior = behaviorManager1.GetBehavior(Of SmartPasteBehavior)(gridControl1.MainView)
            smartPasteBehavior.Properties.ItemDescriptions.AddRange(itemsDescriptions)
        End Sub

        Private Sub ViewInvalidValueException(ByVal sender As Object, ByVal e As InvalidValueExceptionEventArgs)
            e.ExceptionMode = XtraEditors.Controls.ExceptionMode.Ignore
        End Sub

        Private Async Sub GridControlProcessKeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
            If e.KeyData = (Keys.Control Or Keys.Shift Or Keys.V) Then
                e.Handled = True
                Dim smartPasteBehavior = behaviorManager1.GetBehavior(Of SmartPasteBehavior)(gridControl1.MainView)
                Await smartPasteBehavior.SmartPasteAsync()
            End If
        End Sub

        Private Async Sub LayoutControlProcessKeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
            If e.KeyData = (Keys.Control Or Keys.Shift Or Keys.V) Then
                e.Handled = True
                Dim smartPasteBehavior = behaviorManager1.GetBehavior(Of SmartPasteBehavior)(layoutControl1)
                Await smartPasteBehavior.SmartPasteAsync()
            End If
        End Sub

        Private Sub simpleButtonCopy_Click(ByVal sender As Object, ByVal e As EventArgs)
            SafeClipboardWin.Instance.SetText(memoEdit1.Text)
        End Sub

        Private Async Sub simpleButtonPaste_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim smartPasteBehavior = behaviorManager1.GetBehavior(Of SmartPasteBehavior)(layoutControl1)
            Await smartPasteBehavior.SmartPasteAsync()
        End Sub

        Private Sub tabPane1_SelectedPageIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            If tabPane1.SelectedPage Is layoutTabNavigationPage Then
                memoEdit1.Text = LayoutCopyText
            ElseIf tabPane1.SelectedPage Is gridTabNavigationPage Then
                memoEdit1.Text = GridCopyText
            Else
                memoEdit1.Text = SchedulerCopyText
            End If
        End Sub

        Private Class CSharpImpl

            <Obsolete("Please refactor calling code to use normal Visual Basic assignment")>
            Shared Function __Assign(Of T)(ByRef target As T, value As T) As T
                target = value
                Return value
            End Function
        End Class
    End Class

    Public Class Record
        Implements INotifyPropertyChanged

        Public Sub New()
        End Sub

        Private idField As Integer

        Public Property ID As Integer
            Get
                Return idField
            End Get

            Set(ByVal value As Integer)
                If idField <> value Then
                    idField = value
                    OnPropertyChanged()
                End If
            End Set
        End Property

        Private text As String

        <DisplayName("Company")>
        Public Property CompanyName As String
            Get
                Return text
            End Get

            Set(ByVal value As String)
                If Not Equals(text, value) Then
                    If String.IsNullOrEmpty(value) Then Throw New Exception()
                    text = value
                    OnPropertyChanged()
                End If
            End Set
        End Property

        Private val As Decimal?

        <DataType(System.ComponentModel.DataAnnotations.DataType.Currency)>
        <DisplayName("Payment")>
        Public Property Amount As Decimal?
            Get
                Return val
            End Get

            Set(ByVal value As Decimal?)
                If val <> value Then
                    val = value
                    OnPropertyChanged()
                End If
            End Set
        End Property

        Private dt As Date

        <DisplayFormat(DataFormatString:="d")>
        Public Property DueDate As Date
            Get
                Return dt
            End Get

            Set(ByVal value As Date)
                If dt <> value Then
                    dt = value
                    OnPropertyChanged()
                End If
            End Set
        End Property

        Private state As Boolean

        Public Property Processed As Boolean
            Get
                Return state
            End Get

            Set(ByVal value As Boolean)
                If state <> value Then
                    state = value
                    OnPropertyChanged()
                End If
            End Set
        End Property

        Public Overrides Function ToString() As String
            Return String.Format("ID = {0}, Text = {1}", ID, CompanyName)
        End Function

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

        Protected Sub OnPropertyChanged(<CallerMemberName> ByVal Optional propertyName As String = "")
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
    End Class

    Public Class SmartPasteDataHelper

        Public Shared companies As String() = New String() {"Hanari Carnes", "Que Delicia", "Romero y tomillo", "Mere Paillarde", "Comércio Mineiro", "Reggiani Caseifici", "Maison Dewey", "Eastern Delights"}

        Public Shared items As String() = New String() {"Office Chairs", "Laptops", "Printer Ink", "Tables", "Paper"}

        Public Shared Function GetData(ByVal count As Integer) As BindingList(Of Record)
            Dim records As BindingList(Of Record) = New BindingList(Of Record)()
            Dim rnd = NonCryptographicRandom.Default
            For i As Integer = 0 To count - 1
                Dim n As Integer = rnd.Next(10)
                Dim record = New Record() With {.ID = i + 100, .CompanyName = companies(i Mod companies.Length), .DueDate = Date.Today.AddDays(n - 5), .Amount = If(i Mod 2 = 0, (i + 1) * 123, i * 231), .Processed = i Mod 2 = 0}
                records.Add(record)
            Next

            Return records
        End Function
    End Class
End Namespace
